### ISSUE TITLE
[💡 Feature]: Conditional retries

### ISSUE THREAD

**Issue by doctorwait** [webdriverio/webdriverio#11855]
### Is your feature request related to a problem?

Thoughts about such feature starts since I had to work with a huge e2e script (which I can't make smaller). I had "stale element exception" / "stale element error" in a different points of execution that script so I can't set some wrapper for it in a certain place of code - I'm sure there is no such a place. 
The only way for me is to re-run the whole script by setting default num of re-run tries or by side custom Powershell script integrated at CI pipeline. But since execution fires automatically after building back-end in CI pipeline I can't guarantee everything will be OK. And for other test developers it will be nice option to have conditional retries right from the box.

### Describe the solution you'd like.

Today we have an option to set just the number of retries in case of failed test.
Tomorrow we'll have an option to set not just a number but additional some condition for failures so WDIO itself can decide will test execution finally fail by the right reason (e.g. missed assert) or there was some barrier to execute the whole test like stale element error.
1. You wrote a nice test with asserts and other good stuff.
2. You decided that the test can be re-runned 3 times **only** if the reason of fail was **<choose your option>**.
3. You run the test and enjoy ;)

### Describe alternatives you've considered.

That's not a "magic bullet" for sure. An alternative may be some like-linter-tool that during execution will highlight some weak places at the code that suspectible for stale element error (based on many executions).

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by christian-bromann** [comment]
> I had "stale element exception" / "stale element error" in a different points of execution that script

Can you provide some WebdriverIO logs for this error? WebdriverIO has retries for stale element errors, I am wondering if they persisted.

As mentioned in the issue we currently have a `specFileRetries` option. I can see we extend the configuration with something like `specFileRetryOn` that takes a function with the failure error and the number of retries and the user can decide to either retry the spec file or not by returning a boolean.

Any opinions?

**Comment by doctorwait** [comment]
So after some working time and analyzing stacktraces I mentioned that such exception appears while I use this:
```
async clickJS() {
    const click = async function (elem) {
      await browser.executeAsync(async (element, done) => {
        await element.click();
        done();
      }, elem);
    };
    logger.info(`Click on element ${this._name}`);
    try {
      const elem = await this._getThisElement();
      await click(elem);
    } catch (error) {
      if (error.name === 'StaleElementReferenceError') {
        const elem = await this._getThisElement();
        await click(elem);
      } else {
        throw error;
      }
    }
  }
```
It is impossible to apply simple standard click at few places in the program so I use that custom function.
I can not find any valuable info at the logs except stale element row "reference: stale element reference: stale element not found". Other is just regular files queue starts like at...
So maybe there is no huge need in new feature implementation? :)

**Comment by christian-bromann** [comment]
Sure, you can always make a custom implementation of a command and have it conditionally retry. That would need to be done for every specific command though. Feel free to close the issue if you don't have interest in this anymore.

**Comment by Piyush-Chaudhary** [comment]
I agree with @doctorwait point about having the option to rerun spec files based on specific errors. For example, there are instances where network issues might prevent a page from loading correctly, leading to an incorrect response. In such cases, it's crucial to have the capability to rerun the spec file only for those specific errors. While I've figured out how to throw particular errors based on conditions, I haven't yet found a way to rerun the spec file only for those errors, without rerunning it for other types of errors or ignoring them.
